be/src/common/config.h

```diff
@@ -287,6 +287,7 @@ CONF_mInt64(max_cumulative_compaction_num_singleton_deltas, "1000");
 // -1 means no limit if enable event_based_compaction_framework, and the max concurrency will be:
 CONF_Int32(cumulative_compaction_num_threads_per_disk, "1");
 // CONF_Int32(cumulative_compaction_write_mbytes_per_sec, "100");
+CONF_mInt64(max_compaction_candidate_num, "40960");
```

Comment:
while we have a `max_compaction_concurrency` config item there, can we set this compaction candidate based on the max_compaction_concurrency, may a linear factor is good enough?

Reply:
This parameter is typically not intended to be exposed to users. The scenario I'm assuming where we would adjust this parameter is when too many candidate_num result in excessive CPU/memory consumption in the priority queue. In such cases, we may need to reduce it. In this scenario, there is no linear relationship with max_compaction_concurrency.

Reply:
so how to define this "too many"? if we can get clear about "how many" is "too many", maybe we can set a proper value without tune the number up and down. My theory is, there are at most N threads doing the compaction work, the queue can have x times N candidates there waiting for the work, more candidates doesn't help with the compaction, just waste of memory. Maybe I misunderstand this part of logic.

Reply:
All candidates added to the priority queue are tablets that need to do compaction. A low score just indicates that its compaction is not as urgent. If it gets kicked out of the queue, it might not trigger compaction for a long time. Therefore, under favorable resource conditions, we prefer not to kick it out of the queue.

Reply:
that's why the priority queue works here, once kicked out by low priority, and then for a long time there is no trigger about the compaction, which means all other jobs are more important than this compaction job. next time it get chance to enter the queue, remember to take the "time to next compaction" as a fact of weight to the priority and adjust it accordingly.

It is contradicted here, for one hand, I don't want to kick out any job from the queue, on the other hand, I don't want the queue to consume too much resource. It should be a compromise between the resource usage and the queue depth, and this manual adjustment of the parameter may not be the best way to achieve this.

After all, I don't see a clear strategy solution for this, just another switch open here for possible tuning.

Reply:
This parameter is undoubtedly a trade-off. If it weren't for resource limitations, we would certainly not want to restrict the queue's length. However, the length of this queue is unquestionably not linearly related to the number of compaction task threads. Since it's just a limiting value, **the goal is to avoid extreme situations where an excessively long queue leads to OOM and excessive CPU consumption**. Therefore, using an enough large, reasonable resource-consuming, and modifiable configuration as the default is a reasonable choice.